runtime/major_gc.c

```diff
@@ -468,6 +458,18 @@ static intnat get_major_slice_work(intnat howmuch) {
   /* calculate how much work to do now */
   if (howmuch == AUTO_TRIGGERED_MAJOR_SLICE ||
       howmuch == GC_CALCULATE_MAJOR_SLICE) {
+    uintnat heap_size = caml_heap_size(dom_st->shared_heap);
+    uintnat heap_words = (double)Wsize_bsize(heap_size);
+    uintnat heap_sweep_words = heap_words;
+
+    /* cap accumulated work todo to 0.3 */
+    intnat limit = (intnat)(0.3 * (heap_sweep_words
+                            + (heap_words * 100 / (100 + caml_percent_free))));
+    if (dom_st->major_work_todo > limit)
+    {
+      dom_st->major_work_todo = limit;
+    }
```

Comment:
You're still capping the accumulated work-to-do (and doing it as a side effect in a function that shouldn't change it, yuck). What I meant in my review, is that `dom_st->major_work_todo` should not be capped at all, but `computed_work` should be capped instead.

Reply:
I agree that making it a side effect of `get_major_slice_work` is a bit unpleasant. 
I must have misinterpreted what we needed to do here. I've revisited the change to cap `computed_work`.

Reply:
ergh, the CI fail for `tests/parallel/join.ml` is certainly unexpected. I'm investigating it.